_posts/2023-01-11-no-builder-in-kotlin.markdown

```diff
@@ -0,0 +1,184 @@
+Most features of Java and its immediate ecosystem including the likes of Lombok have direct or not so direct counterparts.
+So while rewriting a Java project to Kotlin, should you e.g. encounter a Lombok @Data annotation, it's easy to choose a `data class` on Kotlin's side.
+
+**But what about Lombok's @Builder annotation? There seems to be no in-build builder facility in Kotlin. Do we have to implement a Builder for each class manually?**
+
+**TLDR:** No, we don't. Use optional parameters (=parameters with default arguments) in your Kotlin class and use named parameters when invoking it.  
+
+## What is the problem Lombok @Builder is solving?
+
+Let's assume we have a dto containing many properties including `username` and `password`. The class has one constructor to initialize all its properties.
+Invoking this constructor in Java>=10 will look something like this:
+
+```java
+var a = new A(null, null, "a_schmidt", "1234", null, null);
+```
+
+The issues here include:
+- a lot of visual noise (having to write `null` for all the parameters you don't have a value for)
+- are you sure that username and password are on the right position without jumping to the constructor definition?
+The compiler only guarantees that the types are matching, but how do you know that you didn't mix up the order of - in this case - username and password?
+
+## How Lombok @Builder helps
+
+adding a `@Builder` annotation to `A` will allow you to write code like this
+
+```java
+var a = A.builder()
+            .setUsername("a_schmidt")
+            .setPassword("12324")
+            .build();
+```
+As you can see, we no longer have to provide the parameters we don't have values for, and it's immediate apparent that you didn't confuse the order of parameters.
+It's definitely progress.
+
+**Side note: If you're very concerned about securing that you don't mix up usernames and passwords, you can use wrapper types.**
+After all your organisation might give out usernames like "1234" and allow passwords like "a_schmidt". With Lombok's help, this would look like
+
+```java
+@Value
+class Username {
+    private String value;
+}
+
+@Value
+class Password {
+    private String value;
+}
+
+@Data
+class A {
+    private Role role;
+    private String x1;
+    private Username username;
+    private Password password;
+    private String x2;
+    private String x3;
+}
+```
+
+Now even invoking the constructor without a Builder mixed up would fail (at least for username and password), since
+
+```java
+var a = new A(null, null, new Password("1234"), new Username("a_schmidt"), null, null);
+```
+is a compiletime error.
+
+## Why this isn't a problem in Kotlin to begin with
+
+You don't need to employ the Builder in Kotlin because Kotlin provides optional parameters, named parameters and nullability also plays a certain part. 
```

Comment:
optional parameters, named parameters and nullability ->
optional and named parameters and nullability
(es sind zwei aufzählungen in einer)